### DISCUSSION TITLE
How to get a list of node

### DISCUSSION THREAD

**Discussion by murawakimitsuhiro** [ts-graphviz/ts-graphviz#798] [Q&A] [answered]
I would like to list the nodes in my graph.

The methods I have tried are to use the `nodes()` function externally as shown below, or to extend and get them from the inside.
```ts
const g = parse(`digraph {a ->  c}`)
console.debug(g.nodes)
```

```ts
export class CustomGraph extends Digraph {
    constructor() {
        super('G')
        const maP = new Node('A')
        const saM = new Node('B')
        const saP = new Node('C')
        this.addEdge(new Edge([maP, saP]))
        this.addEdge(new Edge([saM, saP]))

        console.debug(this.nodes)
    }
}
```

The above ways did not correctly retrieve the node array.
![image](https://user-images.githubusercontent.com/13833242/161481477-1fac2423-0539-44e2-add0-a3cf062e559a.png)

Is there already a way to get an array of nodes?
(If my code is incorrect)

Or is this not recommended pattern?

Thanks for your help.

**Comment by kamiazya** [comment, marked as answer]
Hi @murawakimitsuhiro 

The graph does not manage the node designated as an edge element as a graph node.

This is because the node specified when adding an edge is just an alias for the node, and it is not possible to distinguish whether it is a node declared outside the graph as shown below.

```dot
graph G {
    a;
    subgraph cluster_hoge {
        a -- b;
    }
}
```

So, in order to do what you want to do, either declare the node in the graph in advance (A) or get the ID of the target node of the edge (B).

See [runkit demo](https://runkit.com/kamiazya/624a8c0470fbdc0008728630)

---

This issue will be closed, but if you need more, please create a new topic in the discussion.

If this helps, I would appreciate it if you could cooperate with my funding in sponsorship.

Thanks

**Comment by kamiazya** [comment]
@all-contributors please add @murawakimitsuhiro  for question.

[a comment by an automated account is not shown]
